### PR TITLE
Add tool to suggest thresholds for reducing alert noise

### PR DESCRIPTION
**Pull request by Hobbsee**

### CODE REVIEW

File: lib/influx.rb [hunk 1]
```diff
@@ -182,6 +183,44 @@ def noise_candidates(start_date = nil, end_date = nil)
...
+      data.reject! { |x| x['percentile'].nil? || x['percentile'].to_i > recover_within || x['count'] < opts[:more_than].to_i || x['incident_key'].nil? }
+
+      if opts[:sort_by] == 'frequency'
+        data = data.sort_by { |x| x['count'] }.reverse
+      else
+        data = data.sort_by { |x| x[opts[:sort_by]] }
+      end
+
+      data.map do |d|
+        threshold = d['percentile'] + 5
```
Comment (Hobbsee) [inline lib/influx.rb]: This is a static 5 seconds more than the nth percentile alert.  Perhaps this should be a percentage.

File: README.md [hunk 2]
```diff
@@ -9,17 +9,19 @@ At present it offers a graph of alert frequency, but we're also looking to and a
 Pigeonhole requires the following:
 
   - InfluxDB
-  - Ruby 2.1.5 (and rbenv, or your preferred ruby management tool)
+  - Ruby 2.2.1 (and rbenv, or your preferred ruby management tool)
```
Comment (auxesis) [inline README.md]: `s/ruby/Ruby/`

File: bin/alert_threshold_recommendations.rb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+require 'time'
+require 'methadone'
+require 'influx'
+
+include Methadone::Main
+include Methadone::CLILogging
+
+main do
+  raise "Sort by must be one of frequency, percentile, or incident_key" unless %w(frequency percentile incident_key).include?(options[:s])
+  influxdb    = Influx::Db.new
```
Comment (auxesis) [inline bin/alert_threshold_recommendations.rb]: Should remove excess whitespace here: 

``` ruby
influxdb = Influx::Db.new
```

File: lib/influx.rb [hunk 4]
```diff
@@ -182,6 +183,44 @@ def noise_candidates(start_date = nil, end_date = nil)
...
+        formatted_threshold = if threshold < 60
+          "#{threshold} seconds"
+        elsif threshold < 120
```
Comment (auxesis) [inline lib/influx.rb]: `elsif` is pretty much an anti pattern in Ruby. 

This should be refactored to a `case` with multiple `when`s.

Comment (jamtur01) [inline lib/influx.rb]: Also feels like some memoization might be a good idea for the method?

File: lib/influx.rb [hunk 5]
```diff
@@ -182,6 +183,44 @@ def noise_candidates(start_date = nil, end_date = nil)
...
+      # Also remove alerts that don't have an incident key, or haven't been resolved yet.
+      recover_within = ChronicDuration.parse(opts[:recover_within]).to_i
+      raise "Failed to parse recover within" unless recover_within > 0
+      data.reject! { |x| x['percentile'].nil? || x['percentile'].to_i > recover_within || x['count'] < opts[:more_than].to_i || x['incident_key'].nil? }
+
+      if opts[:sort_by] == 'frequency'
+        data = data.sort_by { |x| x['count'] }.reverse
+      else
+        data = data.sort_by { |x| x[opts[:sort_by]] }
+      end
```
Comment (auxesis) [inline lib/influx.rb]: This whole stanza can be shortened: 

``` ruby
sort_by = opts[:sort_by] == 'frequency' ? 'count' : opts[:sort_by]
data = data.sort_by {|x| x[sort_by] }
data.reverse! if sort_by == 'frequency'
```

**Comment by auxesis** [comment]
Other than the stylistic notes above, it looks pretty solid to me, @Hobbsee. :+1:

File: lib/influx.rb (hunk 4 above)
Comment (Hobbsee) [inline lib/influx.rb]: > Also feels like some memoization might be a good idea for the method?

I'm not sure what you mean - could you please elaborate?

File: lib/influx.rb (hunk 5 above)
Comment (Hobbsee) [inline lib/influx.rb]: Great catch, thanks!

File: lib/influx.rb (hunk 4 above)
Comment (jamtur01) [inline lib/influx.rb]: Bah - ignore me. I misread what this did. Nothing to see here. :)

Comment (auxesis) [inline lib/influx.rb]: :hear_no_evil: :see_no_evil: :speak_no_evil:

Comment (Hobbsee) [inline lib/influx.rb]: Wow, those are some great icons!

File: README.md [hunk 6]
```diff
@@ -38,6 +40,12 @@ Options:
                                      (Default: info)
 ```
 
+#### Listen for data from Pagerduty Webhooks
```
Comment (Hobbsee) [inline README.md]: I appear to have written documentation for the wrong branch.

File: bin/alert_threshold_recommendations.rb [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+use_log_level_option
+
+description "Calculates suggested alert thresholds for removing X% of alerts over a given time period"
+
+on("-p percent", "--percent-to-remove", "The percent of alerts to remove")
+on("-t duration", "--time-period", "The amount of time we should look back")
+on("-m more-than", "--more-than", "Only return results that have more than x occurences")
+on("-r recover-within", "--recover-within", "Only return results that recover within this amount of time")
+on("-s sort-by", "--sort-by", "The field we should sort the returned data by - frequency, percentile, or incident_key")
```
Comment (Hobbsee) [inline bin/alert_threshold_recommendations.rb]: I find myself wondering "What's percentile, again?"

I'll rename this to what it is - the time that the threshold should be set to to remove X% of alerts.

File: lib/influx.rb (hunk 4 above)
Comment (auxesis) [inline lib/influx.rb]: http://www.emoji-cheat-sheet.com

**Pull request merged**